File: intellij-lsp/src/com/github/gtache/lsp/contributors/LSPCompletionContributor.scala
package com.github.gtache.lsp.contributors

import com.github.gtache.lsp.PluginMain
import com.github.gtache.lsp.utils.Utils
import com.intellij.codeInsight.completion.{CompletionContributor, CompletionParameters, CompletionResultSet}
import com.intellij.openapi.diagnostic.Logger

/**
  * The completion contributor for the LSP
  */
class LSPCompletionContributor extends CompletionContributor {
  private val LOG: Logger = Logger.getInstance(this.getClass)

  override def fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet): Unit = {
    val editor = parameters.getEditor
    val offset = parameters.getOffset
    val serverPos = Utils.logicalToLSPPos(editor.offsetToLogicalPosition(offset))
    val toAdd = PluginMain.completion(editor, serverPos)
    result.addAllElements(toAdd)
    super.fillCompletionVariants(parameters, result)
  }
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/contributors/LSPCompletionContributor.kt
package com.github.gtache.lsp.contributors

import com.github.gtache.lsp.editor.services.application.EditorApplicationService
import com.github.gtache.lsp.utils.DocumentUtils
import com.intellij.codeInsight.completion.CompletionContributor
import com.intellij.codeInsight.completion.CompletionParameters
import com.intellij.codeInsight.completion.CompletionResultSet
import com.intellij.openapi.components.service

/**
 * LSP completion contributor
 */
class LSPCompletionContributor : CompletionContributor() {

    override fun fillCompletionVariants(parameters: CompletionParameters, result: CompletionResultSet): Unit {
        val editor = parameters.editor
        val offset = parameters.offset
        val serverPos = DocumentUtils.offsetToLSPPosition(editor, offset)
        val toAdd = service<EditorApplicationService>().managerForEditor(editor)?.completion(serverPos) ?: emptyList()
        result.addAllElements(toAdd)
        super.fillCompletionVariants(parameters, result)
    }
}